packages/signature-controller/src/SignatureController.ts

```diff
@@ -0,0 +1,680 @@
+    this._rejectUnapproved(this._personalMessageManager, reason);
+    this._rejectUnapproved(this._typedMessageManager, reason);
+  }
+
+  /**
+   * Clears all unapproved messages from memory.
+   */
+  clearUnapproved() {
+    this._clearUnapproved(this._messageManager);
+    this._clearUnapproved(this._personalMessageManager);
+    this._clearUnapproved(this._typedMessageManager);
+  }
+
+  private _rejectUnapproved<
+    M extends AbstractMessage,
+    P extends AbstractMessageParams,
+    PM extends AbstractMessageParamsMetamask,
+  >(messageManager: AbstractMessageManager<M, P, PM>, reason?: string) {
+    Object.keys(messageManager.getUnapprovedMessages()).forEach((messageId) => {
+      this._cancelAbstractMessage(messageManager, messageId, reason);
+    });
+  }
+
+  private _clearUnapproved<
+    M extends AbstractMessage,
+    P extends AbstractMessageParams,
+    PM extends AbstractMessageParamsMetamask,
+  >(messageManager: AbstractMessageManager<M, P, PM>) {
+    messageManager.update({
+      unapprovedMessages: {},
+      unapprovedMessagesCount: 0,
+    });
+  }
+
+  private async _signAbstractMessage<
+    M extends AbstractMessage,
+    P extends AbstractMessageParams,
+    PM extends AbstractMessageParamsMetamask,
+  >(
+    messageManager: AbstractMessageManager<M, P, PM>,
+    methodName: string,
+    msgParams: PM,
+    getSignature: (cleanMessageParams: P) => Promise<any>,
+  ) {
+    console.info(`MetaMaskController - ${methodName}`);
+
+    const messageId = msgParams.metamaskId as string;
+
+    try {
+      const cleanMessageParams = await messageManager.approveMessage(msgParams);
+      const signature = await getSignature(cleanMessageParams);
+
+      messageManager.setMessageStatusSigned(messageId, signature);
+
+      this._acceptApproval(messageId);
+
+      return this._getState();
+    } catch (error: any) {
+      console.info(`MetaMaskController - ${methodName} failed.`, error);
+      this._errorMessage(messageManager, messageId, error.message);
+      throw error;
+    }
+  }
+
+  private _errorMessage<
+    M extends AbstractMessage,
+    P extends AbstractMessageParams,
+    PM extends AbstractMessageParamsMetamask,
+  >(
+    messageManager: AbstractMessageManager<M, P, PM>,
+    messageId: string,
+    error: string,
+  ) {
+    if (messageManager instanceof TypedMessageManager) {
+      messageManager.setMessageStatusErrored(messageId, error);
+      this._rejectApproval(messageId);
+    } else {
+      this._cancelAbstractMessage(messageManager, messageId);
+    }
+  }
+
+  private _cancelAbstractMessage<
+    M extends AbstractMessage,
+    P extends AbstractMessageParams,
+    PM extends AbstractMessageParamsMetamask,
+  >(
+    messageManager: AbstractMessageManager<M, P, PM>,
+    messageId: string,
+    reason?: string,
+  ) {
+    if (reason) {
+      const message = this._getMessage(messageId);
+      this.hub.emit('cancelWithReason', { message, reason });
```

Comment:
Is the plan to add an event listener for this event to the extension and mobile? I cannot find an existing one.

Reply:
Exactly. This was the last thing I did in the extension, so was still testing at the time.